### PR TITLE
`Job.BuildNumberAssigner`

### PR DESCRIPTION
**Pull request by jglick**
In CloudBees CI in high availability mode, the default behavior of `Job.assignBuildNumber` (updating a counter in memory, and writing it to disk) is not suitable for rapid build triggers of a single job; an external counter is needed at least transiently (the `nextBuildNumber` file is fine for a durable count). This beta extension point allows a plugin to swap in an alternate strategy for maintaining the next build number.

~The `BuildNumberAssignment` return type is a bit awkward but I could not think of a better API. `Job.saveNextBuildNumber` is `protected`, so not usable from alternate implementations, and I did not want to make it `public` solely for the benefit of a beta API. (Introducing a `public` alias also seemed weird.) The method introduced in #9019 does not save, whereas the older `updateNextBuildNumber` method remains undesirable here because of its call to `getLastBuild`. Automatically calling `saveNextBuildNumber` from `assignBuildNumber` would change sychronization behavior for the default implementation, and making `assignBuildNumber` be `synchronized` would be undesirable for the CloudBees CI implementation.~ Reworked, I hope more clearly.

This PR makes #9778 not strictly necessary but I think it remains appropriate anyway.

### Testing done

Effectiveness confirmed by a functional test in CloudBees CI.

### Proposed changelog entries

- N/A

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/java/hudson/model/Job.java
Comment (Vlatombe) [inline core/src/main/java/hudson/model/Job.java:379]: The only alternative I can think of would be, but it looks as awkward.
```
@FunctionalInterface
public interface Saver<T> {
   void save(T t) throws IOException;
}
```
and
```
BuildNumberAssignment assignBuildNumber(Job<?, ?> job, Saver<Job<?,?> saver) throws IOException;
```

Comment (jglick) [reply core/src/main/java/hudson/model/Job.java:379]: I guess you mean

```java
int assignBuildNumber(Job<?, ?> job, Saver<Job<?,?>> saver) throws IOException;
```

or more simply (since we only want to allow _this_ job to be saved)

```java
interface Saver {
    void saveNextBuildNumber() throws IOException;
}
int assignBuildNumber(Job<?, ?> job, Saver saver) throws IOException;
```

called as

```java
return assignBuildNumber(this, this::saveNextBuildNumber);
```

Would work. I am not sure if it is any clearer.

Comment (Vlatombe) [reply core/src/main/java/hudson/model/Job.java:379]: Yes. Not vastly clearer but at least the implementation has full control on when save occurs.

**Review by Vlatombe — APPROVED** [review]
LGTM

**Review by timja — APPROVED** [review]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by MarkEWaite** [comment]
Plugin BOM reports a test failure in the jacoco plugin when testing weekly with Jenkins 2.484.

https://github.com/jenkinsci/bom/pull/3912#issuecomment-2458115654

**Pull request merged**
